File: src/main/kotlin/model/Forest.kt
package model

internal class Forest {
    private var predecessor: HashMap<String, String> = HashMap()
    private var rank: HashMap<String, Int> = HashMap()

    internal fun makeSet(x: String) {
        predecessor[x] = x
        rank[x] = 0
    }

    /**
     * Merges the root nodes of `recessiveNode`
     * and `dominantNode` into one node, where
     * `dominantNode` will be the new root node.
     * Thus, the key/name of `dominantNode`'s
     * root node will be kept.
     *
     * @param recessiveNode Node to merge, will omit its name.
     * @param dominantNode  Node to merge, will keep its name.
     */
    internal fun union(recessiveNode: String, dominantNode: String) {
        link(findSet(recessiveNode)!!, findSet(dominantNode)!!)
    }

    /**
     * Links `recessiveNode` and `dominantNode` into one node,
     * where `dominantNode` will be the new root node.
     * Thus, the key/name of `dominantNode` will be kept.
     *
     * @param recessiveNode Node to link, will omit its name.
     * @param dominantNode  Node to link, will keep its name.
     */
    private fun link(recessiveNode: String, dominantNode: String) {
        if (rank[recessiveNode]!! > rank[dominantNode]!!) {
            predecessor[dominantNode] = recessiveNode
            return
        }

        predecessor[recessiveNode] = dominantNode

        if (rank[recessiveNode] == rank[dominantNode])
            rank[dominantNode] = rank[dominantNode]!!.plus(1)
    }

    internal fun findSet(x: String): String? {
        if (x != predecessor[x])
            predecessor[x] = findSet(predecessor[x]!!)!!

        return predecessor[x]
    }

    override fun toString(): String {
        return "Forest(predecessor=$predecessor, rank=$rank)"
    }
}

File: src/lib/model/forest.ts
export class Forest {
    private _predecessor: { [key: string]: string; } = {};
    private _rank: { [key: string]: number; } = {};

    public get predecessor(): { [key: string]: string; } {
        return this._predecessor;
    }

    public get rank(): { [key: string]: number; } {
        return this._rank;
    }

    public makeSet(x: string): void {
        this._predecessor[x] = x;
        this._rank[x] = 0;
    }

    /**
     * Merges the root nodes of `recessiveNode`
     * and `dominantNode` into one node, where
     * `dominantNode` will be the new root node.
     * Thus, the key/name of `dominantNode`'s
     * root node will be kept if union by rank is not violated.
     *
     * @param recessiveNode Node to merge, will omit its name.
     * @param dominantNode  Node to merge, will keep its name.
     */
    public union(recessiveNode: string, dominantNode: string): void {
        this.link(this.findSet(recessiveNode)!, this.findSet(dominantNode)!);
    }

    /**
     * Links `recessiveNode` and `dominantNode` into one node,
     * where `dominantNode` will be the new root node.
     * Thus, the key/name of `dominantNode` will be kept.
     *
     * But: Union by rank will always win over `dominantNode`.
     *
     * @param recessiveNode Node to link, will omit its name.
     * @param dominantNode  Node to link, will keep its name.
     */
    private link(recessiveNode: string, dominantNode: string): void {
        if (this._rank[recessiveNode]! > this._rank[dominantNode]!) {
            this._predecessor[dominantNode] = recessiveNode;
            return;
        }

        this._predecessor[recessiveNode] = dominantNode;

        if (this._rank[recessiveNode] === this._rank[dominantNode]) {
            this._rank[dominantNode] = this._rank[dominantNode]! + 1;
        }
    }

    /**
     * Finds the root node of a tree.
     * 
     * @param x     An element of a tree.
     * @returns     Root node of the tree containing `x`, or `undefined` if not found.
     */
    public findSet(x: string): string | undefined {
        // Path compression
        if (x !== this._predecessor[x]) {
            this._predecessor[x] = this.findSet(this._predecessor[x]!)!;
        }

        return this._predecessor[x];
    }
}